File: src/main/java/de/espend/idea/php/annotation/doctrine/action/DoctrineClassGeneratorAction.java
package de.espend.idea.php.annotation.doctrine.action;

import com.intellij.codeInsight.CodeInsightActionHandler;
import com.intellij.codeInsight.actions.CodeInsightAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.util.PsiTreeUtil;
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocComment;
import com.jetbrains.php.lang.psi.PhpFile;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import de.espend.idea.php.annotation.dict.PhpDocCommentAnnotation;
import de.espend.idea.php.annotation.doctrine.util.DoctrineUtil;
import de.espend.idea.php.annotation.util.AnnotationUtil;
import org.jetbrains.annotations.NotNull;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
abstract public class DoctrineClassGeneratorAction extends CodeInsightAction {
    @Override
    protected boolean isValidForFile(@NotNull Project project, @NotNull Editor editor, @NotNull PsiFile file) {
        if (!(file instanceof PhpFile) || !DoctrineUtil.isDoctrineOrmInVendor(project)) {
            return false;
        }

        int offset = editor.getCaretModel().getOffset();
        if (offset <= 0) {
            return false;
        }

        PsiElement psiElement = file.findElementAt(offset);
        if (psiElement == null) {
            return false;
        }

        if (!PlatformPatterns.psiElement().inside(PhpClass.class).accepts(psiElement)) {
            return false;
        }

        PhpClass phpClass = PsiTreeUtil.getParentOfType(psiElement, PhpClass.class);
        if (phpClass == null) {
            return false;
        }

        PhpDocComment docComment = phpClass.getDocComment();
        if (docComment == null) {
            return true;
        }

        PhpDocCommentAnnotation container = AnnotationUtil.getPhpDocCommentAnnotationContainer(docComment);

        return container == null || container.getPhpDocBlock(supportedClass()) == null;
    }

    @NotNull
    @Override
    protected CodeInsightActionHandler getHandler() {
        return new CodeInsightActionHandler() {
            @Override
            public void invoke(@NotNull Project project, @NotNull Editor editor, @NotNull PsiFile file) {

                int offset = editor.getCaretModel().getOffset();
                if(offset <= 0) {
                    return;
                }

                PsiElement psiElement = file.findElementAt(offset);
                if(psiElement == null) {
                    return;
                }

                if(!PlatformPatterns.psiElement().inside(PhpClass.class).accepts(psiElement)) {
                    return;
                }

                PhpClass phpClass = PsiTreeUtil.getParentOfType(psiElement, PhpClass.class);
                if(phpClass == null) {
                    return;
                }

                // insert ORM alias
                execute(editor, phpClass, file);
            }

            @Override
            public boolean startInWriteAction() {
                return true;
            }
        };
    }

    abstract protected void execute(@NotNull Editor editor, @NotNull PhpClass phpClass, @NotNull PsiFile psiFile);

    /**
     * Class supported by this action. if already inside DocBlock dont provide an action
     */
    @NotNull
    abstract protected String supportedClass();
}

File: src/main/kotlin/de/espend/idea/php/annotation/doctrine/action/DoctrineClassGeneratorAction.kt
package de.espend.idea.php.annotation.doctrine.action

import com.intellij.codeInsight.CodeInsightActionHandler
import com.intellij.codeInsight.actions.CodeInsightAction
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.patterns.ElementPattern
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.util.PsiTreeUtil
import com.jetbrains.php.lang.psi.PhpFile
import com.jetbrains.php.lang.psi.elements.PhpClass
import de.espend.idea.php.annotation.doctrine.util.DoctrineUtil
import de.espend.idea.php.annotation.util.AnnotationUtil
import org.apache.commons.lang3.StringUtils

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
abstract class DoctrineClassGeneratorAction : CodeInsightAction() {
    override fun isValidForFile(project: Project, editor: Editor, file: PsiFile): Boolean {
        if (file !is PhpFile || !DoctrineUtil.isDoctrineOrmInVendor(project)) {
            return false
        }

        val offset = editor.caretModel.offset
        if (offset <= 0) {
            return false
        }

        val psiElement = file.findElementAt(offset) ?: return false
        if (!INSIDE_PHP_CLASS_PATTERN.accepts(psiElement)) {
            return false
        }

        val phpClass = PsiTreeUtil.getParentOfType(psiElement, PhpClass::class.java) ?: return false
        val docComment = phpClass.docComment
        if (docComment != null) {
            val container = AnnotationUtil.getPhpDocCommentAnnotationContainer(docComment)
            if (container?.getPhpDocBlock(supportedClass()) != null) {
                return false
            }
        }

        val supportedClass = "\\" + StringUtils.stripStart(supportedClass(), "\\")
        return phpClass.getAttributes(supportedClass).isEmpty()
    }

    override fun getHandler(): CodeInsightActionHandler {
        return object : CodeInsightActionHandler {
            override fun invoke(project: Project, editor: Editor, file: PsiFile) {
                val offset = editor.caretModel.offset
                if (offset <= 0) {
                    return
                }

                val psiElement = file.findElementAt(offset) ?: return
                if (!INSIDE_PHP_CLASS_PATTERN.accepts(psiElement)) {
                    return
                }

                val phpClass = PsiTreeUtil.getParentOfType(psiElement, PhpClass::class.java) ?: return
                execute(editor, phpClass, file)
            }

            override fun startInWriteAction(): Boolean = true
        }
    }

    protected abstract fun execute(editor: Editor, phpClass: PhpClass, psiFile: PsiFile)

    /**
     * Class supported by this action. If it is already inside the DocBlock, do not provide an action.
     */
    protected abstract fun supportedClass(): String

    private companion object {
        val INSIDE_PHP_CLASS_PATTERN: ElementPattern<PsiElement> =
            PlatformPatterns.psiElement().inside(PhpClass::class.java)
    }
}
